Prepare SSSD to support IPA in trust to Samba AD

abbra opened the pull request:
This pull request prepares SSSD ipa provider to support IPA in trust to Samba AD but the same changes are needed for a properly working bi-directional trust against Microsoft AD as well. To make everything fully working, one needs patches against FreeIPA too but SSSD changes are isolated.

@sumit-bose @jhrozek please review.

1. When IPA establishes a trust to an Active Directory forest, a number of special objects is created in a subtree of `cn=trusts,$SUFFIX`. These objects represent Kerberos principals for trusted domain objects (TDOs) used for both incoming and outgoing trusts. For bi-directional trust there is a requirement that one of them (`<REMOTE FLAT NAME>$@<OUR REALM>`) must have a POSIX identity because a remote domain controller will use it to authenticate against smbd running on IPA master.

SSSD only looks for user accounts in `cn=accounts,$SUFFIX`, so an attempt by smbd to resolve this principal name as a POSIX user via `getpwnam()` will fail. And the reason why smbd behaves this way is due to the fact that a Kerberos ticket used for authentication contains no MS-PAC record, thus not allowing Samba to build a local security token it needs. This is expected for the authentication using TDO account as it is used for bootstrapping reasons (AD DC couldn't create and sign MS-PAC record for an account in IPA realm) but the side effect is that TDO object must be known as a POSIX account on IPA master.

Thus, we extend user search base in IPA provider to search in both `cn=accounts,$SUFFIX` and `cn=trusts,$SUFFIX`. Changes on FreeIPA side will handle access controls and generation of the POSIX information for the TDO accounts.

2. For long time we relied on using cross-realm TGTs to talk to Active Directory domain controllers (LDAP and GC services) in case of bi-directional trust. Unfortunately, this is not something we can continue using as there are multiple reasons such access can be denied by a trusted AD side, including SID filtering and other security measurements. It also happens that right now Samba AD in Fedora has a bug in handling a cross-realm TGT generated by the FreeIPA KDC. As result, while technically IPA could establish a bi-directional trust to Samba AD, it does not work as any SSSD attempt to connect to AD DCs via LDAP with GSSAPI will fail (Samba AD DC answers error with PROCESS_TGS message on Kerberos level and authentication fails).

For this reason, we should remove any distinction when using bi-directional trust and simply always use a special keytab with a TDO object as we do in uni-directional trust case. While a more generic Kerberos authentication will not work in the outbound direction, SSSD will be able to resolve users/groups.

sumit-bose commented:
ok to test

jhrozek commented:
I only scrolled through the patches, I must admit I didn't do any testing yet.

For the search bases, I would just appreciate a more descriptive variable names than value/value2. 

Is the TDO lookup something that also the clients do or only masters do? The first patch touches both cases, but I think it might improve performance if we only set multiple search bases on the IPA servers. In general I'm somewhat concerned about performance implications because two search bases mean that anything that doesn't hit the first search base also hits the second one. And in the typical case, this might be either local users (because of the way libc behaves, it tries to find group memberships across all databases) or, in case of trusts, we first try to match a name against the ID views before the trusted domain, now we'd have to try for the ID view in two bases. This is OK if it's done only on the master, but if this was done on the clients as well, I wonder if we had to improve the heuristics or add a filter to the search base.

btw is adding the filter something we should do anyway to speed up the lookup? In this case, we do know what the server is, so we might as well ask the 389 developers if they have any recommendation.

About the keytab patch, this really needs careful testing. At the very least, we also have codepaths for one-way or two-way trusts that also set different options (I guess that's where the principal and the keytab get propagated into the low-level sdap code). Did you test the lookups with both one-way and two-way trusts after the change or would you prefer if I (or someone else, see below) try to run the available tests?

finally please note I'll be mostly away much of the next week -- I hope others can chime in as well if this PR is urgent, otherwise I'll see what I can do.

abbra commented:
@jhrozek @sumit-bose I've updated the patches to address Jakub's comments.TDO lookup should only be performed in the server mode, so I added that. At the server side there will be ACIs limiting this access to `cn=adtrust agents` members (e.g. SSSD and smbd on IPA masters) only. I also added a filter `(objectclass=ipaIDObject)` which will be used by the TDO objects.

I'm going to test it more extensively today.

jhrozek commented:
I'm sorry about the delay @abbra I was on PTO last week.

Code-wise, the code looks OK, but I'm afraid I hit a bug in my testing:
```
(Tue Mar  6 10:44:37 2018) [sssd[be[ipa.test]]] [ipa_get_id_options] (0x0400): Option ldap_user_search_base set to cn=accounts,dc=ipa,dc=test?cn=trusts,dc=ipa,dc=test??(objectclass=ipaIDObject)
(Tue Mar  6 10:44:37 2018) [sssd[be[ipa.test]]] [common_parse_search_base] (0x0020): Unparseable search base: [cn=accounts,dc=ipa,dc=test?cn=trusts,dc=ipa,dc=test??(objectclass=ipaIDObject)][4]
(Tue Mar  6 10:44:37 2018) [sssd[be[ipa.test]]] [ipa_init_id_ctx] (0x0020): Unable to init id context [22]: Invalid argument
(Tue Mar  6 10:44:37 2018) [sssd[be[ipa.test]]] [sssm_ipa_init] (0x0020): Unable to init IPA ID context [22]: Invalid argument
(Tue Mar  6 10:44:37 2018) [sssd[be[ipa.test]]] [dp_module_run_constructor] (0x0010): Module [ipa] constructor failed [22]: Invalid argumen
```

abbra commented:
I think there could be two approaches here:
- either parse existing base to see if there are multiple bases or parameters already and re-assemble with a new one
- or directly add a new base after parsing an existing one

I'd love to avoid re-assembling the bases but there is little we could do to check how many parameters are there in the existing base. E.g. [cn=accounts,dc=ipa,dc=test] or [cn=account,dc=ipa,dc=test?subtree?(|(objectClass=posixAccount)(objectClass=posixGroup))] or [cn=account,dc=ipa,dc=test??(|(objectClass=posixAccount)(objectClass=posixGroup))] -- if how many '?' should I add without parsing? There is clearly an issue in SSSD itself in how to submit these bases.

I'd rather have a function where I can pass a current array of bases+parameters and a new base+parameters and get back a resulting string. I think this would be useful for others too.

abbra commented:
@jhrozek ping, is it possible to review this pull request?

jhrozek commented:
coverity and CI did not find any issues -> +Accepted

jhrozek commented:
* master:
14faec9cd9437ef116ae054412d25ec2e820e409
0648053a7c99de5148c4cffea68c47a3f660303d
